File: src/dissectors/beacon_dissector.rs
use crate::utils::TypeConverter;


pub(crate) struct BeaconDissector;


impl BeaconDissector {

    pub fn parse_beacon(beacon: &[u8]) -> Option<Vec<String>> {
        let frame = Self::skip_headers(beacon);
        
        if frame.len() < 24 {
            return None;
        }

        let frame_ctrl    = u16::from_le_bytes([frame[0], frame[1]]);
        let frame_type    = (frame_ctrl >> 2) & 0x03; 
        let frame_subtype = (frame_ctrl >> 4) & 0x0F;

        if frame_type != 0 || frame_subtype != 8 {
            return None;
        }

        let bssid = Self::get_bssid(frame);
        let ssid  = Self::get_ssid(frame);
        let chnl  = Self::get_channel(frame);
        let sec   = Self::get_sec_type(frame);

        Some(vec![ssid, bssid, chnl.to_string(), sec])
    }



    fn skip_headers(beacon: &[u8]) -> &[u8] {
        if let Some(offset) = Self::find_frame_start_by_type(beacon) {
            return &beacon[offset..];
        }

        if let Some(offset) = Self::skip_radiotap_header(beacon) {
            return &beacon[offset..];
        }

        if let Some(offset) = Self::skip_common_headers(beacon) {
            return &beacon[offset..];
        }

        beacon
    }



    fn find_frame_start_by_type(beacon: &[u8]) -> Option<usize> {
        for i in 0..beacon.len().saturating_sub(24) {
            if i + 2 > beacon.len() {
                continue;
            }

            let frame_ctrl = u16::from_le_bytes([beacon[i], beacon[i+1]]);
            let frame_type    = (frame_ctrl >> 2) & 0x03;
            let frame_subtype = (frame_ctrl >> 4) & 0x0F;

            if frame_type != 0 || frame_subtype != 8 {
                continue;
            }

            if i + 3 >= beacon.len() {
                continue;
            }

            let duration = u16::from_le_bytes([beacon[i+2], beacon[i+3]]);
            if duration > 0x3AFF {
                continue;
            }

            return Some(i);
        }

        None
    }



    fn skip_radiotap_header(beacon: &[u8]) -> Option<usize> {
        if beacon.len() < 8 || beacon[0] != 0x00 || beacon[1] != 0x00 {
            return None
        }

        let radiotap_len = u16::from_le_bytes([beacon[2], beacon[3]]) as usize;
        if radiotap_len < 8 || radiotap_len + 24 > beacon.len() {
            return None;
        }

        let after_radiotap = &beacon[radiotap_len..];
        if after_radiotap.len() < 24 {
            return None;
        }

        let frame_ctrl = u16::from_le_bytes([after_radiotap[0], after_radiotap[1]]);
        let frame_type = (frame_ctrl >> 2) & 0x03;
        if frame_type > 2 {
            return None;
        }

        Some(radiotap_len)
    }



    fn skip_common_headers(beacon: &[u8]) -> Option<usize> {
        let common_offsets = [0, 4, 8, 12, 16, 24, 32, 36];

        for &offset in &common_offsets {
            if offset + 24 > beacon.len() {
                continue;
            }

            let frame_data    = &beacon[offset..];
            let frame_ctrl    = u16::from_le_bytes([frame_data[0], frame_data[1]]);
            let frame_type    = (frame_ctrl >> 2) & 0x03;
            let frame_subtype = (frame_ctrl >> 4) & 0x0F;
            
            if frame_type != 0 || frame_subtype != 8 {
                continue;
            }

            return Some(offset);
        }

        None
    }



    fn get_bssid(frame: &[u8]) -> String {
        if frame.len() < 22 {
            return "00:00:00:00:00:00".to_string();
        }

        let bssid_bytes = &frame[16..22];
        TypeConverter::mac_vec_u8_to_string(bssid_bytes)
    }



    fn get_ssid(frame: &[u8]) -> String {
        if frame.len() < 38 {
            return "<hidden>".to_string();
        }

        let mut offset = 36;

        while offset + 1 < frame.len() {
            let element_id     = frame[offset];
            let element_length = frame[offset + 1] as usize;
            offset += 2;

            if element_id != 0 || element_length <= 0 {                
                offset += element_length;
                continue;
            }

            let ssid_start = offset;
            let ssid_end   = ssid_start + element_length;

            if ssid_end > frame.len() {
                offset += element_length;
                continue;
            }

            let ssid_bytes = &frame[ssid_start..ssid_end];
                    
            if ssid_bytes.iter().all(|&b| b == 0) {
                return "<hidden>".to_string();
            }
    
            if let Ok(ssid) = String::from_utf8(ssid_bytes.to_vec()) {
                if !ssid.trim().is_empty() {
                    return ssid;
                }
            }
                    
            return Self::format_ssid_bytes(ssid_bytes);
        }

        "<hidden>".to_string()
    }



    fn format_ssid_bytes(ssid_bytes: &[u8]) -> String {
        let is_printable = ssid_bytes.iter().all(|&b| b >= 32 && b <= 126);
        
        if is_printable {
            return String::from_utf8_lossy(ssid_bytes).to_string();
        }
        
        let display_len      = std::cmp::min(8, ssid_bytes.len());
        let hex_part: String = ssid_bytes[..display_len]
            .iter()
            .map(|b| format!("{:02X}", b))
            .collect::<Vec<String>>()
            .join("");
            
        if ssid_bytes.len() > display_len {
            return format!("{}...", hex_part);
        } 
        
        hex_part
    }



    fn get_channel(frame: &[u8]) -> u8 {
        if frame.len() < 38 {
            return 0;
        }

        let mut offset = 36;

        while offset + 1 < frame.len() {
            let element_id     = frame[offset];
            let element_length = frame[offset + 1] as usize;
            offset += 2;

            if element_id != 3
               || element_length != 1
               || offset >= frame.len()
            {
                offset += element_length;
                continue;
            }

            return frame[offset];
        }

        0
    }



    fn get_sec_type(frame: &[u8]) -> String {
        if frame.len() < 38 {
            return "????".to_string();
        }

        let mut sec_flags = SecFlags::new();
        let mut offset = 36;

        while offset + 1 < frame.len() {
            let element_id = frame[offset];
            let element_length = frame[offset + 1] as usize;

            if offset + 2 + element_length > frame.len() {
                break;
            }

            let element_data = &frame[offset + 2..offset + 2 + element_length];

            match element_id {
                0x30 => Self::process_rsn_element(element_data, &mut sec_flags),
                0xDD => Self::process_vendor_element(element_data, &mut sec_flags),
                0x06 => Self::process_privacy_element(element_data, &mut sec_flags),
                _ => {}
            }

            offset += 2 + element_length;
        }

        sec_flags.to_sec_string()
    }



    fn process_rsn_element(data: &[u8], flags: &mut SecFlags) {
        flags.has_rsn = true;
        flags.is_open = false;

        if data.len() >= 20 {
            flags.is_wpa3 = Self::check_for_wpa3(data);
        }
    }



    fn check_for_wpa3(rsn_data: &[u8]) -> bool {
        let mut offset = 6; // Version (2 bytes) Group Cipher Suite (4 bytes)

        if offset + 2 > rsn_data.len() {
            return false;
        }

        let pairwise_count = u16::from_le_bytes([rsn_data[offset], rsn_data[offset + 1]]) as usize;
        offset += 2 + (pairwise_count * 4);

        if offset + 2 > rsn_data.len() {
            return false;
        }

        let akm_count = u16::from_le_bytes([rsn_data[offset], rsn_data[offset + 1]]) as usize;
        offset += 2;

        for _ in 0..akm_count {
            if offset + 3 >= rsn_data.len() {
                break;
            }

            if rsn_data[offset] != 0x00 ||
               rsn_data[offset + 1] != 0x0F ||
               rsn_data[offset + 2] != 0xAC {
                offset += 4;
                continue;
            }

            let suite_type = rsn_data[offset + 3];
            if suite_type == 8 || suite_type == 9 {
                return true;
            }

            offset += 4;
        }

        false
    }



    fn process_vendor_element(data: &[u8], flags: &mut SecFlags) {
        if data.len() >= 4 &&
           data[0] == 0x00 &&
           data[1] == 0x50 &&
           data[2] == 0xF2 &&
           data[3] == 0x01 {
            flags.has_wpa = true;
            flags.is_open = false;
        }
    }



    fn process_privacy_element(data: &[u8], flags: &mut SecFlags) {
        if !data.is_empty() && data[0] & 0x10 != 0 {
            flags.has_wep = true;
            flags.is_open = false;
        }
    }

}



struct SecFlags {
    has_rsn: bool,
    has_wpa: bool,
    has_wep: bool,
    is_open: bool,
    is_wpa3: bool,
}

impl SecFlags {
    fn new() -> Self {
        SecFlags {
            has_rsn: false,
            has_wpa: false,
            has_wep: false,
            is_open: true,
            is_wpa3: false,
        }
    }

    fn to_sec_string(&self) -> String {
        if self.is_wpa3 {
            "WPA3".to_string()
        } else if self.has_rsn {
            "WPA2".to_string()
        } else if self.has_wpa {
            "WPA".to_string()
        } else if self.has_wep {
            "WEP".to_string()
        } else if self.is_open {
            "Open".to_string()
        } else {
            "Unknown".to_string()
        }
    }
}

File: dissectors/beacon_dissector.go
package dissectors

import (
	"encoding/binary"
	"fmt"
	"strings"
	"unicode/utf8"
)



type BeaconDissector struct {
    frame []byte
}



func NewBeaconDissector() *BeaconDissector {
    return &BeaconDissector{}
}



func (bd *BeaconDissector) DissecBeacon(beacon []byte) ([]string, bool) {
    bd.frame = beacon

    offset   := bd.findFrameOffset()
    bd.frame  = beacon[offset:]
    
	if len(bd.frame) < 24 {
        return nil, false
    }
    
	frameCtrl    := binary.LittleEndian.Uint16(bd.frame[0:2])
    frameType    := (frameCtrl >> 2) & 0x03
    frameSubtype := (frameCtrl >> 4) & 0x0F
    
	if frameType != 0 || frameSubtype != 8 {
        return nil, false
    }
    
	bssid   := bd.getBSSID()
    ssid    := bd.getSSID()
    channel := bd.getChannel()
    sec     := bd.getSecType()
    
	return []string{ssid, bssid, channel, sec}, true
}



func (bd *BeaconDissector) findFrameOffset() int {
    if offset, ok := bd.findFrameStartByType(); ok {
        return offset
    }

	if offset, ok := bd.skipRadiotapHeader(); ok {
        return offset
    }

	if offset, ok := bd.skipCommonHeaders(); ok {
        return offset
    }

	return 0
}



func (bd *BeaconDissector) findFrameStartByType() (int, bool) {
    for i := 0; i+24 <= len(bd.frame); i++ {

		if i+2 > len(bd.frame) {
            continue
        }

		frameCtrl    := binary.LittleEndian.Uint16(bd.frame[i : i+2])
        frameType    := (frameCtrl >> 2) & 0x03
        frameSubtype := (frameCtrl >> 4) & 0x0F

		if frameType != 0 || frameSubtype != 8 {
            continue
        }

		if i+4 > len(bd.frame) {
            continue
        }
        
		duration := binary.LittleEndian.Uint16(bd.frame[i+2 : i+4])
        if duration > 0x3AFF {
            continue
		}
    
		return i, true
    }
    
	return 0, false
}



func (bd *BeaconDissector) skipRadiotapHeader() (int, bool) {
    if len(bd.frame) < 8 || bd.frame[0] != 0x00 || bd.frame[1] != 0x00 {
        return 0, false
    }

	radiotapLen := int(binary.LittleEndian.Uint16(bd.frame[2:4]))
    if radiotapLen < 8 || radiotapLen+24 > len(bd.frame) {
        return 0, false
    }

	afterRadiotap := bd.frame[radiotapLen:]
    if len(afterRadiotap) < 24 {
        return 0, false
    }

	frameCtrl := binary.LittleEndian.Uint16(afterRadiotap[0:2])
    frameType := (frameCtrl >> 2) & 0x03

	if frameType > 2 {
        return 0, false
    }

	return radiotapLen, true
}



func (bd *BeaconDissector) skipCommonHeaders() (int, bool) {
    commonOffsets := []int{0, 4, 8, 12, 16, 24, 32, 36}
    
	for _, offset := range commonOffsets {
        if offset+24 > len(bd.frame) {
            continue
        }

		frame        := bd.frame[offset:]
        frameCtrl    := binary.LittleEndian.Uint16(frame[0:2])
        frameType    := (frameCtrl >> 2) & 0x03
        frameSubtype := (frameCtrl >> 4) & 0x0F
 
		if frameType == 0 && frameSubtype == 8 {
            return offset, true
        }
    }
    
	return 0, false
}



func (bd *BeaconDissector) getBSSID() string {
    if len(bd.frame) < 22 {
        return "00:00:00:00:00:00"
    }

	b := bd.frame[16:22]
    return fmt.Sprintf("%02x:%02x:%02x:%02x:%02x:%02x", b[0], b[1], b[2], b[3], b[4], b[5])
}



func (bd *BeaconDissector) getSSID() string {
    const hidden = "<hidden>"

	if len(bd.frame) < 38 {
        return hidden
    }

	offset := 36
    for offset+1 < len(bd.frame) {
        elementID  := bd.frame[offset]
        elementLen := int(bd.frame[offset+1])
        offset += 2
        
		if elementID != 0 || elementLen <= 0 {
            offset += elementLen
            continue
        }
        
		if offset+elementLen > len(bd.frame) {
            offset += elementLen
            continue
        }

		ssidBytes := bd.frame[offset : offset+elementLen]
        allZero   := true

		for _, b := range ssidBytes {
            if b != 0 {
                allZero = false
                break
            }
        }
        
		if allZero {
            return hidden
        }

		if utf8.Valid(ssidBytes) {
            ssid := string(ssidBytes)
            if strings.TrimSpace(ssid) != "" {
                return ssid
            }
        }

		return formatSSIDBytes(ssidBytes)
    }

	return hidden
}



func formatSSIDBytes(ssid []byte) string {
    isPrintable := true

	for _, b := range ssid {
        if b < 32 || b > 126 {
            isPrintable = false
            break
        }
    }

	if isPrintable {
        return string(ssid)
    }

	displayLen := 8
    if len(ssid) < displayLen {
        displayLen = len(ssid)
    }

	hexPart := ""
    for i := 0; i < displayLen; i++ {
        hexPart += fmt.Sprintf("%02X", ssid[i])
    }

	if len(ssid) > displayLen {
        return hexPart + "..."
    }

	return hexPart
}



func (bd *BeaconDissector) getChannel() string {
    if len(bd.frame) < 38 {
        return "0"
    }

	offset := 36
    for offset+1 < len(bd.frame) {
        elementID  := bd.frame[offset]
        elementLen := int(bd.frame[offset+1])
        offset += 2
        
		if elementID == 3 && elementLen == 1 && offset < len(bd.frame) {
            return fmt.Sprintf("%d", bd.frame[offset])
        }
        
		offset += elementLen
    }
    
	return "0"
}



func (bd *BeaconDissector) getSecType() string {
    if len(bd.frame) < 38 {
        return "????"
    }

	flags  := &secFlags{isOpen: true}
    offset := 36

	for offset+1 < len(bd.frame) {
        elementID  := bd.frame[offset]
        elementLen := int(bd.frame[offset+1])
        
		if offset+2+elementLen > len(bd.frame) {
            break
        }
        
		elementData := bd.frame[offset+2 : offset+2+elementLen]
        
		switch elementID {
        case 0x30: processRSNElement(elementData, flags)
        case 0xDD: processVendorElement(elementData, flags)
        case 0x06: processPrivacyElement(elementData, flags)
        }
        
		offset += 2 + elementLen
    }
    
	return flags.String()
}



type secFlags struct {
	hasRSN  bool
	hasWPA  bool
	hasWEP  bool
	isOpen  bool
	isWPA3  bool
}



func processRSNElement(data []byte, flags *secFlags) {
	flags.hasRSN = true
	flags.isOpen = false

	if len(data) >= 20 {
		flags.isWPA3 = checkForWPA3(data)
	}
}



func checkForWPA3(rsnData []byte) bool {
	offset := 6

	if offset+2 > len(rsnData) {
		return false
	}
	
	pairwiseCount := int(binary.LittleEndian.Uint16(rsnData[offset : offset+2]))
	offset += 2 + pairwiseCount*4
	
	if offset+2 > len(rsnData) {
		return false
	}
	
	akmCount := int(binary.LittleEndian.Uint16(rsnData[offset : offset+2]))
	offset += 2
	
	for i := 0; i < akmCount; i++ {
		if offset+4 > len(rsnData) {
			break
		}

		if rsnData[offset] == 0x00 && rsnData[offset+1] == 0x0F && rsnData[offset+2] == 0xAC {
			suiteType := rsnData[offset+3]
			if suiteType == 8 || suiteType == 9 {
				return true
			}
		}

		offset += 4
	}

	return false
}



func processVendorElement(data []byte, flags *secFlags) {
	if len(data) >= 4 && data[0] == 0x00 && data[1] == 0x50 && data[2] == 0xF2 && data[3] == 0x01 {
		flags.hasWPA = true
		flags.isOpen = false
	}
}



func processPrivacyElement(data []byte, flags *secFlags) {
	if len(data) > 0 && (data[0]&0x10) != 0 {
		flags.hasWEP = true
		flags.isOpen = false
	}
}



func (f *secFlags) String() string {
	switch {
	case f.isWPA3: return "WPA3"
	case f.hasRSN: return "WPA2"
	case f.hasWPA: return "WPA"
	case f.hasWEP: return "WEP"
	case f.isOpen: return "Open"
	default:       return "Unknown"
	}
}